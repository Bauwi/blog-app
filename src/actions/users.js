import { db } from '../firebase/firebase';

export const updateUser = (id, updates) => ({
  type: 'UPDATE_USER',
  id,
  updates
});

export const startUpdateUser = (id, updates) => (dispatch, getState) => {
  const uid = id || getState().auth.uid;
  return db
    .collection('users')
    .doc(uid)
    .get()
    .then((doc) => {
      if (doc.exists) {
        return db
          .collection('users')
          .doc(uid)
          .update(updates)
          .then(() => {
            dispatch(updateUser(uid, updates));
          });
      }
      return db
        .collection('users')
        .doc(uid)
        .set({
          avatar: 'http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png',
          email: '',
          username: '',
          stars: 0,
          topCategories: ['life', 'programming', 'Music']
        })
        .then(startUpdateUser(uid, updates));
    });
};

export const addUserStar = nextStars => ({
  type: 'ADD_USER_STAR',
  nextStars
});

export const startAddUserStar = (id, prevStars) => dispatch =>
  db
    .collection('users')
    .doc(id)
    .update({ stars: prevStars + 1 })
    .then(() => dispatch(addUserStar(prevStars + 1)));

export const setAuthor = author => ({
  type: 'SET_AUTHOR',
  author
});

export const startSetAuthor = id => dispatch =>
  db
    .collection('users')
    .doc(id)
    .get()
    .then((doc) => {
      dispatch(setAuthor(doc.data()));
    });

export const setUserPreferences = preferences => ({
  type: 'SET_USER_PREFERENCES',
  preferences
});

export const startSetUserPreferences = preferences => (dispatch, getState) =>
  db
    .collection('users')
    .doc(getState().auth.uid)
    .get()
    .then((doc) => {
      dispatch(setUserPreferences(doc.data()));
    });
