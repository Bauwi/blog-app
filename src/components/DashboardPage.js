import React from 'react';

import PostsList from './PostsList';

export const DashboardPage = () => (
  <div>
    <PostsList />
  </div>
);

export default DashboardPage;
