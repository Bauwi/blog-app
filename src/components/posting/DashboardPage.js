import React from 'react';

import PostsList from './PostsList';
import PostsSummary from './PostsSummary';

export const DashboardPage = () => (
  <div>
    <PostsSummary />
    <PostsList />
  </div>
);

export default DashboardPage;
