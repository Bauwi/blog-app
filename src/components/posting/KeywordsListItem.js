import React from 'react';

const KeywordsListItem = ({ keyword }) => <li>{keyword}, </li>;

export default KeywordsListItem;
