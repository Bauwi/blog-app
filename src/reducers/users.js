const usersReducerDefaultState = [];

const usersReducer = (state = usersReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_USER':
      return {
        ...state,
        preferences: action.newUser
      };
    case 'UPDATE_USER_PREFERENCES':
      return {
        ...state,
        preferences: { ...state.preferences, ...action.updates }
      };
    case 'SET_AUTHOR':
      return {
        ...state,
        author: action.author
      };
    case 'ADD_USER_STAR':
      return {
        ...state,
        author: {
          ...state.author,
          stars: action.nextStars
        }
      };
    case 'SET_USER_PREFERENCES':
      return {
        ...state,
        preferences: action.preferences
      };
    case 'USERS_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'USERS_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    default:
      return state;
  }
};

export default usersReducer;
