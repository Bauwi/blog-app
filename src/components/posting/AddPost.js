import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PostForm from './PostForm';
import { startAddPost } from '../../actions/posts';
import { startUpdateUser } from '../../actions/users';

export class AddPost extends Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.startUpdateUser(undefined, { username: post.author });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <header className="content-container">
            <Link to="/dashboard">Back to dashboard</Link>
          </header>
        </div>
        <PostForm onSubmit={this.onSubmit} context="add" author={this.props.user} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post)),
  startUpdateUser: (id, updates) => dispatch(startUpdateUser(id, updates))
});

const mapStateToProps = state => ({
  user: state.users.preferences
});
export default connect(mapStateToProps, mapDispatchToProps)(AddPost);
