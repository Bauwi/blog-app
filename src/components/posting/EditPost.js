import React, { Component } from 'react';
import { connect } from 'react-redux';
import { startEditPost, startRemovePost } from '../../actions/posts';

import PostForm from './PostForm';

export class EditPost extends Component {
  onSubmit = post => {
    this.props.startEditPost(this.props.post.id, post);
    this.props.history.push('/dashboard');
  };
  onRemovePost = () => {
    this.props.startRemovePost(this.props.post.id);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <PostForm onSubmit={this.onSubmit} post={this.props.post} />
        <button onClick={this.onRemovePost}>Remove Post</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    post: state.posts.find(post => {
      return post.id === props.match.params.id;
    })
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditPost: (id, post) => dispatch(startEditPost(id, post)),
  startRemovePost: id => dispatch(startRemovePost(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPost);
