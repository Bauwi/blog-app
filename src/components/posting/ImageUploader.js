import React, { Component } from 'react';
import { connect } from 'react-redux';
import { uploadImage } from '../../actions/posts';

export class ImageUploader extends Component {
  handleChange = e => {
    const image = e.target.files[0];
    this.props.uploadImage(image);
  };

  render() {
    return (
      <div>
        <progress value="0" max="100">
          0%
        </progress>
        <input type="file" onChange={this.handleChange} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  uploadImage: image => dispatch(uploadImage(image))
});

export default connect(undefined, mapDispatchToProps)(ImageUploader);
