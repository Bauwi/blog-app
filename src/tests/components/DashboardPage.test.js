import React from 'react';
import { shallow } from 'enzyme';
import { DashboardPage } from '../../components/posting/DashboardPage';
import { user1 } from '../fixtures/users';

test('should render DashboardPage component', () => {
  const wrapper = shallow(<DashboardPage preferences={user1} userId="123abc" />);
  expect(wrapper).toMatchSnapshot();
});
