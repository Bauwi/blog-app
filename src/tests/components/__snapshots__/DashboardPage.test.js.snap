// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DashboardPage component 1`] = `
<div>
  <Link
    replace={false}
    to="/create"
  >
    New Post
  </Link>
  <Connect(PostsList) />
</div>
`;
